auto_update.go

```diff
@@ -0,0 +1,163 @@
+package main
+
+import (
+	"archive/zip"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"runtime"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/inconshreveable/go-update"
+)
+
+const locallyBuilt = "(Locally Built)"
+
+// RunWithUpdateCheck checks if an update is due, checks if current version is outdated and performs update if needed
+func (c *TGFConfig) RunWithUpdateCheck() int {
+	app := c.tgf
+	const autoUpdateFile = "TGFAutoUpdate"
+
+	if !app.AutoUpdateSet { // Fallback to config settings if --update flag is not set
```

Comment:
The logic here is really weird.
Setting app.AutoUpdateSet doesn't make much sense. This variable indicates that the user has set the `--update` or `--no-update` flag. Modifying it can lead to unexpected behavior. In general, modifying global variables to do local work is a big no-no

Here, you don't have to modify any variables to do what you want to do
```
if app.AutoUpdateSet {
	if app.AutoUpdate {
		app.Debug("Auto update is forced. Checking version...")
	} else {
		app.Debug("Auto update is force disabled. Bypassing update version check.")
		return c.Run()
	}
} else {
	if c.AutoUpdate {
		app.Debug("Auto update is disabled in the config. Bypassing update version check.")
		return c.Run()
	}
	if !lastRefresh(autoUpdateFile) >= c.AutoUpdateDelay {
		app.Debug("Less than 2 hours since last check. Bypassing update version check.")
		return c.Run()
	}
}

touchImageRefresh(autoUpdateFile)
...

```